source/common/formatter/substitution_formatter.cc

```diff
@@ -49,6 +49,9 @@ const std::regex& getStartTimeNewlinePattern() {
 }
 const std::regex& getNewlinePattern() { CONSTRUCT_ON_FIRST_USE(std::regex, "\n"); }
 
+template <class... Ts> struct JsonFormatMapVisitor : Ts... { using Ts::operator()...; };
+template <class... Ts> JsonFormatMapVisitor(Ts...) -> JsonFormatMapVisitor<Ts...>;
```

Comment:
Is this second template used? If I read it correctly, it's defining a templated method that takes `Ts...` as parameters and returns a `JsonFormatMapVisitor<Ts...>` but it's not implemented anywhere and in `toStruct` you seem to just use the struct template directly.

Reply:
Without the second one C++ can't deduce the template arguments when initializing the struct, so the alternative is to explicitly list the arguments which I thought looks quite ugly given how complex the callback types are.

Reply:
Without it the compiler can't deduce the template arguments, so the alternative is to list the types, which I thought is quite ugly given how large the type definitions for the lamas are.